### PR TITLE
[NVPTX] Generalize and extend upsizing when lowering 8/16-bit-element vector loads/stores

### PR DESCRIPTION
**Pull request by dakersnar**
This addresses the following issue I opened: https://github.com/llvm/llvm-project/issues/118851.

This change generalizes the Type Legalization mechanism that currently handles `v8[i/f/bf]16` upsizing to include loads _and_ stores of `v8i8` + `v16i8`, allowing all of the mentioned vectors to be lowered to ptx as vectors of `b32`. This extension also allows us to remove the DagCombine that only handled exactly `load v16i8`, thus centralizing all the upsizing logic into one place.

Test changes include adding v8i8, v16i8, and v8i16 cases to load-store.ll, and updating the CHECKs for other tests to match the improved codegen.

### CODE REVIEW

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (AlexMaclean) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6325]: This looks like a full copy paste of a similar switch above, can these be extracted into a small helper function?

**Review by AlexMaclean — COMMENTED** [review]
Nice, this seems like an overall improvement, barring some minor nits.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1407]: This could use some editing. The key limitation is that PTX syntax only supports v2/v4. We can't use vectorized loads/stores with the actual element type for i8/i16 as that would need v8/v16 variants that do not exist. In order to load/store such vectors efficiently, we perform the loads/stores as vectors of i32.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1408]: We could extract the check into a helper function/lanbda with a meaningful name. The function can carry the explanation comment, so we don't have to replicate it everywhere.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3246]: This can be generalized to 
```
NPerWord = 32/EltVT.getSizeInBits(); // elements to pack in one word.
EltVT = getVectorVT(EltVT, NPerWord); // word-sized vector.
NumElts = NumElts/NPerWord; // number of word-sized vectors.
```

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:6325]: Yup. A good candidate for a helper function or lambda.

File: llvm/test/CodeGen/NVPTX/LoadStoreVectorizer.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/LoadStoreVectorizer.ll:182]: I thought we've switched to PRMT for all sorts of byte extractions. Is this patch based on top-of-the-tree or some older snapshot? Yup, looks like it's behind a bit. You may want to update your tree, as this test output may change.

File: llvm/test/CodeGen/NVPTX/load-store.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/load-store.ll:214]: This looks like another optimization opportunity. When we need to extract 4 i8 values, into i16 it may be faster to do this way:

Let's assume input `I = i32 XXYYZZWW`. We need to produce four `i16` values `XX/YY/ZZ/WW`.

```
    PRMT %ZW, %I, 0, 0x4140;   // ZW = 0x00ZZ00WW 
    PRMT %XY, %I, 0, 0x4342;  // XY = 0x00XX00WW
    mov.b32 {%rsZ, %rsW}, %ZW
    mov.b32 {%rsX, %rsY}, %XY
```

On GPUs that support v2i16 operations, we could do them directly on `%ZW` and `%XY`.

This should probably go into a separate patch.

File: llvm/test/CodeGen/NVPTX/load-store.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/load-store.ll:7]: This file has grown to be pretty large. Perhaps we should consider splitting some of the tests into a separate file.

**Review by Artem-B — COMMENTED** [review]
Nice. Thank you for the patch.
LGTM overall, with a few minor suggestions.

File: llvm/test/CodeGen/NVPTX/LoadStoreVectorizer.ll
Comment (dakersnar) [reply llvm/test/CodeGen/NVPTX/LoadStoreVectorizer.ll:182]: Ahh, yeah that might be possible, I can regenerate it soon.

File: llvm/test/CodeGen/NVPTX/load-store.ll
Comment (dakersnar) [reply llvm/test/CodeGen/NVPTX/load-store.ll:7]: That seems reasonable. Any suggestions for how to split it?

**Comment by dakersnar** [comment]
Great suggestions, thanks all! I'll try to implement these tomorrow.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3251]: Good suggestion. Would you still recommend hoisting this to a helper function now that we've reduced it to three lines?

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp
Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1408]: Should this helper live in NVPTXISelDAGToDAG.cpp, or in NVPTXUtilities.cpp like `Is2x16VT` does?

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3251]: I guess. Keeping the magic in one place is a good thing. The helper could return `std::pair(NumElts, EltVT)` and structured binding lets us do it with virtually no boilerplate overhead.
```
auto helper = [](int N, EVT VT) {
  return std::pair(newN), newVT);
};
...
auto [N, VT] = helper(oldN, oldVT);
```

File: llvm/test/CodeGen/NVPTX/load-store.ll
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/load-store.ll:7]: Scalars vs vectors would be one factor.
If vectors test is still too large, it could be split by element type/size (e.g. native vectors vs the ones with small elements that get lowered as b32)

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1408]: Depends on where it's used. If it's needed in multiple files, I'd put them into utils. If it's used only in one file, keep it there.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3251]: Not super familiar with this syntax, do you know how to get around this issue? The problem is that we want to reassign the same variables that we pass into the lambda, rather than create new ones.
image
image

File: llvm/test/CodeGen/NVPTX/LoadStoreVectorizer.ll
Comment (dakersnar) [reply llvm/test/CodeGen/NVPTX/LoadStoreVectorizer.ll:182]: I'm not seeing any difference in the test generation after merging with TOT.

File: llvm/test/CodeGen/NVPTX/load-store.ll
Comment (dakersnar) [reply llvm/test/CodeGen/NVPTX/load-store.ll]: Done. I tried to keep only the relevant TODOs at the top of each new file, can you double check my work there?

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: With my most recent push, I believe I resolved everything except for this. I have a working solution with the lambda but it isn't very clean, if you have suggestions I would love to hear them.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: I think this is resolved, and I also updated ReplaceLoadVector to also use ExtractVectorElements. Let me know if they both look good now.

File: llvm/test/CodeGen/NVPTX/load-store.ll
Comment (dakersnar) [reply llvm/test/CodeGen/NVPTX/load-store.ll]: Good idea. And especially since this affects the existing v4i8 lowering, I agree that this would be better in a separate patch. Do you want to open a tracking bug?

**Comment by dakersnar** [comment]
I just noticed and patched an edge case I had missed: ld.global.nc gets Selected via a different path, and the code was written with the assumption that v4i8 would never be a subvector of a larger vector. I pushed the simplest fix with my most recent commit, plus the test cases that were crashing, but I can also refactor the handling in tryLDGLDU to match what we do in tryLoadVector if you all think it would be a good idea. We could even have it generate `b32` instead of `u32` if we want to be consistent, although maybe that's too much churn.

Any other edge cases you can think of that I may have missed? I can scan through every use of v4i8 in the backend to see if anything seems equally fragile.

**Comment by dakersnar** [comment]
I'm seeing evidence that `Isv2x16VT(VT) || VT == MVT::v4i8` is already being checked at quite a few places. We may want to generalize the helper function `isVectorElementTypeUpsized` helper I just created and use it everywhere. Name suggestions appreciated.

File: llvm/test/CodeGen/NVPTX/LoadStoreVectorizer.ll
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/LoadStoreVectorizer.ll:182]: OK. Looks like we may have more opportunities for switching to PRMT, but it's outside of the scope of this patch.

File: llvm/test/CodeGen/NVPTX/ldg-invariant.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/ldg-invariant.ll:59]: We should convert this test to auto-generated checks, too.

File: llvm/test/CodeGen/NVPTX/shuffle-vec-undef-init.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/shuffle-vec-undef-init.ll:14]: Interesting. We seem to be loading twice as much data as we did before.
While the generated code is valid, and in practice will work faster (old code would effectively fetch the same data, due to caching, only using twice as many instructions), I'm somewhat concerned that we're now loading unused data.

I think we could previously eliminate unused loads because the vector was split, but when we signed up for loading the whole vector, we have no choice but load all elements.

That may become problematic in corner cases when we need to use only few elements of the vector.
We may need to eventually take into account whether individual vector elements are used or not, and fall back to splitting the vector if only few are used.

This may be worth adding a TODO at the point where we make that decision.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1587]: It should also have an assert that NumElts is a multiple of 4.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:103]: Given that it's not inside of the function, this does not have to be a lambda. A function would do.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3216]: This whole switch should probably be folded into `GetUpsizedNumEltsAndEltVT` and `UpsizeElementTypes` returned as an additional `bool`.
Then we could just use `auto [Upsize, NumElts, EltVT] = GetUpsizedNumEltsAndEltVT(ValVT.getVectorNumElements(), ValVT.getVectorElementType())`

Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3216]: Hmm, do you think that will hurt readability? I think having the Upsize helper compute NumElts and EltVT for all (including non-upsized) cases is a bit more confusing.

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3216]: I look at it as "Calculate the right vector shape for lowering", where the "right shape" may be identical to the input shape.

Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3216]: Got it. That sounds reasonable, although we would also need to return a "Native" boolean, right? Returning early if it is false?

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3216]: Or it could be derived from EltVT != original EltVT.

File: llvm/test/CodeGen/NVPTX/shuffle-vec-undef-init.ll
Comment (dakersnar) [reply llvm/test/CodeGen/NVPTX/shuffle-vec-undef-init.ll:14]: Yeah I noticed this and came to the same conclusions.  Like you said, overall this is a reduction in total number of load instructions, so I think it is a net win.

As for possible corner cases where it wouldn't be a win, my gut says it is unlikely that we will find many end-to-end cases that arrive at the backend with a substantial percentage of unused loads. DCE should run a few times before the Vectorizer, right, so unless something happens _after_ the vectorizer runs that makes those loads become unused, I think we can generally trust that vector loads handed to the backend should be lowered as given whenever possible.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3216]: I think there's a misunderstanding; I was referring to cases that are neither native _nor_ upsize-able. This switch statement has two functions right now, marking the cases that we need to upsize, and returning early for cases that cannot be handled at all, such as `v8i32`. How will we handle the latter with the helper function approach, besides making it a "Try" function and adding another boolean to the returns?

Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3216]: I implemented the above std::tie solution but let me know if you still think there is a better way. I'm still hesitant to fold the entire switch into the helper function because the switch is responsible for exiting early from `ReplaceLoadVector`/`LowerSTOREVector` and I don't think there is a clean way to do it, but if I'm missing something I'm happy to tweak it.

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3216]: Fair enough. So, the answers we're looking for are: should we handle this vector, and if so, how.

That asks for either returning a `std::optional<std::pair<>>` or just represent "can't do it" as `NumElts == 0`.
The latter is probably more concise. The function entry would look roughly like this:
```
auto NumElts, EltType = getVectorLoweringParams(Val);
if (NumElts == 0)
  return SDValue();

... lower as `NumElts x EltType`.
```

File: llvm/test/CodeGen/NVPTX/shuffle-vec-undef-init.ll
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/shuffle-vec-undef-init.ll:14]: SGTM. We'll cross that bridge if/when we get there.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3239]: Stype nit: no need for `{}` around single statement bodies.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3216]: std::optional is fine, too.

Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Great, I just pushed the implementation, let me know what you think.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3237]: Nit. Given that we can't assign the variables before checking for the nullopt, perhaps we should return an `optional<struct>` so we can access the values as fields without copying returned values into scalar local vars.

Current code is fine, too.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3272]: We have potentially three outcomes:
* NumElts didn't change. That's supposed to be handled by the `else` clause.
* NumElts decreased. We've upsized vector elements to 32 bits. This is what the true bllock does now.
* NumElts increased. That's an error, and it will currently result in us executing the `else` block, which would be a wrong thing to do.

I think we need an assert before the `if` that we do not expect NumElts to ever increase, or change `else` into a strict equality check and add `else llvm_unreachable()`;

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:3274]: I think what we really care about here is that EltVT is a vector, is 32 bits in size, and has no more than 4 elements. Checking for specific types does the job, but relies on the implementation detail and we have to change it every time we add another type.

**Review by Artem-B — COMMENTED** [review]
Almost there. :-)
Couple more nits and we're done.

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp
Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp]: Trying to resist the urge to scope-creep and refactor more in this function... let me know if what I've done with my most recent push is good enough

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Keeping the current method, I think with only two elements std::pair is fine

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Done, added asserts

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp
Comment (dakersnar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Done, changed assert. Can you double check my API usage to make sure I'm representing these checks correctly?

File: llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp]: LGTM.

**Review by Artem-B — APPROVED** [review]
LGTM. Thank you for the patch and I appreciate your willingness to go through all the changes during the review.

**Comment by dakersnar** [comment]
No problem, the feedback was super helpful, thanks for the detail!

**Comment by dakersnar** [comment]
Just rebased and reran tests, everything looks good. Should be ready to merge. @Artem-B Can you merge this for me? Thank you :)

**Comment by dakersnar** [comment]
Actually, there's one more thing I want to test, hold off for a little I'll let you know when it's ready.

**Comment by Artem-B** [comment]
Also, just in case, please do run LLVM tests with ptxas enabled. Set `LLVM_PTXAS_EXECUTABLE=<PATH_TO_CUDA>/bin/ptxas` before running the tests and will also check if ptxas is happy with the instructions llc generates during the tests. Sometimes it catches issues that slip through LLVM-only tests.

**Comment by dakersnar** [comment]
Done and done. I think Alex is taking a quick look, but otherwise this should be good to go.

**Review by AlexMaclean — APPROVED** [review]
LGTM. merging on @dakersnar's behalf per his offline request.

**Comment by dakersnar** [comment]
Fix for the above errors merged here: https://github.com/llvm/llvm-project/pull/120336

**Pull request merged**
